clients/slo/slo.go

```diff
@@ -0,0 +1,213 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slo
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/clients/slo"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/rest"
+)
+
+func NewClient(client *rest.Client) *Client {
+	c := &Client{
+		client: slo.NewClient(client),
+	}
+	return c
+}
+
+type Client struct {
+	client client
+}
+
+//go:generate mockgen -source slo.go -package=slo -destination=client_mock.go
+type client interface {
+	List(ctx context.Context, ro rest.RequestOptions) (*http.Response, error)
+	Get(ctx context.Context, id string, ro rest.RequestOptions) (*http.Response, error)
+	Create(ctx context.Context, data []byte, ro rest.RequestOptions) (*http.Response, error)
+	Update(ctx context.Context, id string, optimisticLockingVersion string, data []byte, ro rest.RequestOptions) (*http.Response, error)
+	Delete(ctx context.Context, id string, ro rest.RequestOptions) (*http.Response, error)
+}
+
+var _ client = (*slo.Client)(nil)
+
+const errMsgWithId = "failed to %s slo resource with id %s: %w"
+
+// List gets a complete set of complete configuration for all available SLOs
+func (c Client) List(ctx context.Context) (api.PagedListResponse, error) {
+	var retVal api.PagedListResponse
+
+	for haveNextPage, nextPageKey := true, ""; haveNextPage; {
+		response, err := c.client.List(ctx, rest.RequestOptions{
+			CustomShouldRetryFunc: rest.RetryIfTooManyRequests,
+			QueryParams:           url.Values{"page-key": {nextPageKey}}})
+		defer closeBody(response)
+		if err != nil {
+			return nil, err
+		}
+
+		var listResponse api.ListResponse
+		nextPageKey, listResponse, err = processListResponse(response, unmarshallFromListResponse)
+		if err != nil {
+			return nil, err
+		}
+
+		retVal = append(retVal, listResponse)
+		haveNextPage = nextPageKey != ""
+	}
+
+	return retVal, nil
+}
+
+// Get gets a complete configuration of SLO with an ID
+func (c Client) Get(ctx context.Context, id string) (api.Response, error) {
+	resp, err := c.client.Get(ctx, id, rest.RequestOptions{CustomShouldRetryFunc: rest.RetryIfTooManyRequests})
+	defer closeBody(resp)
+	if err != nil {
+		return api.Response{}, fmt.Errorf(errMsgWithId, "get", id, err)
+	}
+	return processResponse(resp, nil)
+}
+
+func (c Client) Create(ctx context.Context, body json.RawMessage) (api.Response, error) {
+	resp, err := c.client.Create(ctx, body, rest.RequestOptions{CustomShouldRetryFunc: rest.RetryIfTooManyRequests})
+	defer closeBody(resp)
+	if err != nil {
+		return api.Response{}, err // error message from the client is descriptive enough
+	}
+	return processResponse(resp, nil)
+}
+
+func (c Client) Update(ctx context.Context, id string, body json.RawMessage) (api.Response, error) {
+	getResp, err := c.Get(ctx, id)
```

Comment:
`getResp` body is not closed.

Reply:
Here it's not a `http.Response`. This is a call to the Get method of this client, and `http.Response` is handled there.